src/gravel/controllers/resources/network.py

```diff
@@ -0,0 +1,328 @@
+    cls: str = Field(title="Device class")
+    vendor: str = Field(title="Device driver's vendor")
+    device: str = Field(title="Device driver's name")
+    hw_vendor: str = Field(title="Hardware vendor")
+    hw_device: str = Field(title="Hardware device")
+
+
+class NetworkDeviceModel(BaseModel):
+    address: str = Field(title="Network device's physical address")
+    is_up: bool = Field(False, title="Network device is UP")
+    speed: int = Field(0, title="Network device operating speed (Mbps)")
+
+
+class NetworkConfigModel(BaseModel):
+    address: Optional[str] = Field(None, title="IP Address")
+    prefixlen: Optional[int] = Field(None, title="Prefix length of IP address")
+    bootproto: Optional[str] = Field(
+        None, title="Mode the interface will be set up"
+    )
+    startmode: Optional[str] = Field(
+        None, title="When the interface should be set up"
+    )
+
+
+class NetworkInterfaceModel(BaseModel):
+    name: str = Field(title="Network Interface name")
+    device: NetworkDeviceModel = Field(title="Network Device Information")
+    config: NetworkConfigModel = Field(title="Network Interface Config")
+    info: PCIModel = Field(title="PCI Information")
+
+
+class NetworkInterfaceStatsModel(BaseModel):
+    tx: int = Field(0, title="Transmitted bytes")
+    rx: int = Field(0, title="Received bytes")
+    ts: int = Field(0, title="Timestamp")
+
+
+class NetworkInterfaces(Ticker):
+
+    # these need to be strings, and be instantiated as Path's on usage, so
+    # we can properly test this with pyfakefs. Otherwise pathlib.Path() will
+    # actually be pointing to the real path, and we won't be able to fake the
+    # file system operations on these files.
+    devpath: str = "/sys/bus/pci/devices"
+    netpath: str = "/sys/class/net"
+    sysconf: str = "/etc/sysconfig/network"
+
+    """ Network Interface mapping, by interface name """
+    _interface_by_name: Dict[str, NetworkInterfaceModel] = {}
+    """ Statistics history, by interface name, in bytes per second """
+    _statistics_by_name: Dict[str, List[NetworkInterfaceStatsModel]] = {}
+    """ Last obtained statistics for a given interface, raw bytes """
+    _last_statistics_by_name: Dict[str, NetworkInterfaceStatsModel] = {}
+
+    def __init__(self, interval: float):
+        super().__init__(interval)
+
+    async def _do_tick(self) -> None:
+        await self._update_interfaces()
+        await self._cleanup_stats()
+        await self._update_stats()
+
+    async def _should_tick(self) -> bool:
+        return True
+
+    def _get_devices(self) -> List[str]:
+        """ Obtain all PCI devices from /sys/bus/pci """
+        devpath = Path(self.devpath)
+        assert devpath.exists() and devpath.is_dir()
+        return next(os.walk(self.devpath))[1]
+
+    async def _get_pci_info(self, pcidev: str) -> PCIModel:
+        """ Obtain specific PCI device information, using lspci. """
+        cmd = f"lspci -vmm -s {pcidev}"
```

Comment:
Why not using [pyudev](https://pyudev.readthedocs.io/en/latest/) instead of running CLI commands and parsing their output?

Reply:
because I wasn't aware of it -- thanks. Regardless, I think the path forward is relying on `lshw` because it provides a lot more information in one go about the network interfaces.